### ISSUE TITLE
Store arbitrary data inside Atoms types

### ISSUE THREAD

**Issue by jameskermode** [libAtoms/JuLIP.jl#28]
I would like to be able to store additional per-atom scalar and vector and tensor properties inside the Atoms type. Examples include local energy, masses, atomic numbers, momenta and/or velocities, local stresses.

Would an `at.arrays::Dict{String,Any}` that maps names to values be a reasonable solution? Should `1xN`, `3xN` and `3x3xN` properties live in different places?

**Comment by cortner** [comment]
kind of a duplicate of #2 , but let's discuss it here.  I've been meaning to do this for a while now.

First of all, it should be arbitrary data, not just arrays. 

My thought was that there should be two dicts:  `persistent` and `transient`.
*  `transient` arrays are cleared or deleted whenever the configuration changes, while 
* `persistent` arrays remain.
 E.g. I would store the neighbourlist as `transient` data, so if I compute energy, then forces, then the neighbourlist is not recomputed. But when the positions change, then I automatically clear the neighbourlist.

What do you think?

**Comment by cortner** [comment]
`unsafe_positions` was in part created so that I can compare new and old positions without making copies all the time.

**Comment by jameskermode** [comment]
The distinction between transient and persistent is a bit more subtle, e.g. if the neighbourlist is computed with a cutoff "skin" beyond that of the potential it remains valid for a few steps (until an atom has moved more than skin).

**Comment by cortner** [comment]
well, instead of `persistent` and `transient`, we can "attach" to every data object a "watcher" which takes as a parameter home much the configuration has changed and returns true or false depending on whether or not the array should be deleted. 

`persistent` arrays then just have a trivial watcher that always returns `false`

**Comment by cortner** [comment]
we can have default watchers, but also write custom ones if ever needed.

**Comment by cortner** [comment]
the watcher could just accumulate the changes and when the sum reaches the threshold then the data is deleted. This would prevent us from having to store many copies of the positions.

**Comment by jameskermode** [comment]
The watchers sound good to me. For interoperability with ASE, the ASEAtoms could actually store the extra data in ``at.po[:arrays]`` Python dictionary, if the overhead is not too great.

**Comment by cortner** [comment]
but keep the watchers in a separate dict in Julia?

**Comment by jameskermode** [comment]
Yes, the watchers don't have an analogue in Python.

ASE also has an ``at.info`` dictionary that is useful for storing arbitrary per-configuration data, e.g. time, temperature or energies evaluated with multiple methods.

**Comment by cortner** [comment]
What would the use-case be for storing data in the `po` instead of the Julia object?

**Comment by jameskermode** [comment]
I was thinking it would be nice to pass the Atoms `po` complete with stored data to Python routines, e.g. `ase.io.extzyz.write()` serialises all the extra data in a format readable by quip, Ovito etc. 

But if it's messy or inefficient we can eventually rewrite I/O in Julia.

**Comment by cortner** [comment]
ok fair, we can think about it. Issues I am anticipating are: 
* reading data back and forth between Julia and Python is either copying of the `unsafe` thing. (which may actually be perfectly in practise!) 
* does it mean we need to restrict this data to *arrays* as opposed to *arbitrary data*?

**Comment by jameskermode** [comment]
How about in the abstract type we have prototypes like the following:

```
data(at::AbstractAtoms, key::String) # returns a copy
unsafe_data(at::AbstractAtoms, key::String) # returns a pointer
set_data!(at::AbstractAtoms, key::String, value)
```

The concrete implementation in `ASEAtoms` would store any arrays with shapes `(length(at),)`, `(length(at), 3)`, etc. in `at.po[:arrays]` and other data in `at.po[:info]`, but eventual other implementations would be free to do something different.

**Comment by cortner** [comment]
a couple of things still bug me: 
 * why distinguish arrays from other datatypes
 * need to test whether *arbitrary* datatypes can be stored in a Python dict (probably)
 * where do we store the information when some data needs to be discarded and recomputed?

**Comment by jameskermode** [comment]
> why distinguish arrays from other datatypes

We don't have to on the Julia side, I propose this as an implementation detail of the ASEAtoms to use the existing ``arrays`` and ``info`` storage.

> need to test whether arbitrary datatypes can be stored in a Python dict (probably)

They certainly can in Python.

>where do we store the information when some data needs to be discarded and recomputed?

A separate list of keys to trash on each positions update?

**Comment by cortner** [comment]
>> where do we store the information when some data needs to be discarded and recomputed?

> A separate list of keys to trash on each positions update?

would this be stored in info then?

**Comment by jameskermode** [comment]
Yes, I think so, unless we need to introduce an attribute in the ASEAtoms type directly for efficiency reasons. I suggest we start with storing all data in the two at.po dictonaries and then profile.

**Comment by cortner** [comment]
Last question, then I can get going on this: I really want to use this for the neighbourlist. This is given as a collection of arrays, but they are not `(length(at), dim)` type arrays, so presumable it should be stored in `info`, correct?

**Comment by cortner** [comment]
> I was thinking it would be nice to pass the `Atoms po` complete with stored data to Python routines, e.g. `ase.io.extzyz.write()` serialises all the extra data in a format readable by quip, Ovito etc.

Does this include the `info` or just the `arrays`?

**Comment by jameskermode** [comment]
Here is the specification: https://libatoms.github.io/QUIP/io.html#module-ase.io.extxyz

`arrays` are written as additional columns in the .xyz file, while contents of `info` are written as key/value pairs to the comment line. This means that `info` should only really be used for "small" quantities or the line gets very long - scalars, vectors, 3x3 matrices are all OK.

I don't think we should store the neighbourlist in there. Noam recently proposed a format for serialising bonds to extxyz, with an additional string entry in `arrays`.

```
4
Properties=species:S:1:bonds:S:1
Si 1
Si 0,3
Si _
Si _
```
where _ means no bonds, and 0-based indices (separated by commas) indicate bonds.

Of course this is just for storage and is not efficient for usage.

**Comment by cortner** [comment]
this is getting complicated now - maybe we need to have a personal chat.

**Comment by cortner** [comment]
Still to decide: should transient arrays be stored in Julia or in Python?

**Comment by cortner** [comment]
Just do it . . .

**Comment by jameskermode** [comment]
I think this can be closed following merge of PR #37
